### DISCUSSION TITLE
session.accessToken is undefined

### DISCUSSION THREAD

**Discussion by iamchathu** [nextauthjs/next-auth#913] [Help] [answered]
**Your question**
I'm using Auth0 provider and I have a 3rd Party Backend which is also secured with Auth0.
I have setup the provider as following but when I try to access session.accessToken, it returned as undefined.

**What are you trying to do**
To retrieve accessToken to access the backend which is not with Next.js.

// api/auth/[...nextauth].ts

```ts
import NextAuth from 'next-auth';
import Providers from 'next-auth/providers';

const options = {
  providers: [
    Providers.Auth0({
      clientId: process.env.AUTH0_CLIENT_ID,
      clientSecret: process.env.AUTH0_CLIENT_SECRET,
      domain: process.env.AUTH0_DOMAIN,
    }),
  ],
  secret: process.env.NEXTAUTH_SECRET,
  jwt: {
    encryption: true,
    secret: process.env.NEXTAUTH_JWT_SECRET,
  },
  pages: {
    signIn: '/auth',
  },
};

export default (req, res) => NextAuth(req, res, options);


```

**Feedback**
*Documentation refers to searching through [online documentation](https://next-auth.js.org), code comments and issue history. The example project refers to [next-auth-example](https://github.com/iaincollins/next-auth-example).*

* [ ] Found the documentation helpful
* [ ] Found documentation but was incomplete
* [x] Could not find relevant documentation
* [ ] Found the example project helpful
* [ ] Did not find the example project helpful

**Comment by balazsorban44** [comment]
Could you please provide a bit more information about how you set up Auth0 with `next-auth`?

**Comment by iamchathu** [comment]
@balazsorban44 Updated the question.

**Comment by balazsorban44** [comment]
What do you mean by:
> To retrieve accessToken to access the backend which is not with Next.js.

Right now, `next-auth` is only intended to be used with Next.js.

You can use the `accessToken` both client and server-side though if you use [API Routes](https://nextjs.org/docs/api-routes/introduction).

Please refer to the documentation:
To add an access token to the session:
- https://next-auth.js.org/configuration/callbacks#jwt-callback
- https://next-auth.js.org/configuration/callbacks#session-callback

To access it from client and server-side:
- https://next-auth.js.org/getting-started/client#getsession
- https://next-auth.js.org/getting-started/client#usesession

**Comment by iamchathu** [comment]
I tried to use `useSession` hook but seems I'm not getting accessToken.

I tried to access accessToken as in 


```ts
const IndexPage: NextPage = () => {
  const classes = useStyles();
  const [session, loading] = useSession();

  const login = () => signIn('auth0');

  return (
    <>
      {!session && (
          <b>
            Sign in
          </b>
          <Button onClick={login} color="primary" variant="contained">
            Sign In
          </Button>
        </AuthLayout>
      )}
      {session && (
        <div>
              <p>Session : {session.expires}</p>
              <p>Token : {session.accessToken}</p>
          </div>
      )}
    </>
  );
};

export default IndexPage;

```

but session.accessToken is `undefined`.

**Comment by balazsorban44** [comment]
You have to forward it in the `jwt` and `session` callbacks, please look at the links.

**Reply by TheRealAstoo** [reply]
Hi @balazsorban44, 

I think my question is a bit related to what you said above : 

When i use session from useSession(), it first is undefined and then i get it. 

Any way to get it without going into the undefined value ? Because it causes my pages to reload everytime i change something related to my user. :)

**Reply by iamchathu** [reply]
> You have to forward it in the `jwt` and `session` callbacks, please look at the links.

I have gone though and I just logged the callback arguments I can not see accessToken coming though there too.

**Reply by TheRealAstoo** [reply]
I don't know if it can help you but i've found how to do it : 

using getServerSideProps with getSession :)

**Reply by iamchathu** [reply]
@TheRealAstoo I don't recommend it. As it 
- enable SSR for the page. 
- creates a Serverless function when hosting.

**Reply by TheRealAstoo** [reply]
And what do you recommend then ? 

Because now it works how i want, only refreshing my  count (for the products, in my cart) instead of refreshing everything due to the fact that useSession is in a hook.

**Reply by balazsorban44** [reply]
@iamchathu Using `getServerSideProps` per se is not a problem, although if it is only to get auth work, than it might actually be unnecessary, since getSession() and useSession() gets their value from the same place:
https://next-auth.js.org/getting-started/rest-api#get-apiauthsession

If you could provide a reproduction with a link to the repository, it would be easier to help :slightly_smiling_face: 

You could also try the `debug: true` option, see where the `access_token` gets lost:
https://next-auth.js.org/configuration/options#debug

If you cannot find an access_token absolutely anywhere, the problem might be in your Auth0 configuration

**Reply by TheRealAstoo** [reply]
Soooo, could it be a good practice to put the session in a getServerSideProps in the App page ? 

Because every time it re renders for a reason, {session.user.email} twitches from nothing to its value when i use useSession() in my component. I've put this in my App : 

`<Provider session={pageProps.session}>
      <Component {...pageProps} />
    </Provider>`

**Reply by TheRealAstoo** [reply]
(thanks for the answer by the way, even if you answered to @iamchathu) 

I can show my repo if needed :)

**Reply by balazsorban44** [reply]
`useSession` is inteded to be used if you need client-side access to your session, `getSession` can be used on server-side.

https://nextjs.org/docs/basic-features/data-fetching
https://nextjs.org/docs/advanced-features/custom-app

You can read more about the different data fetching techniques and its caveats on the above links. It all depends on what you need for your app though. Please check out the Next.js documentation.

Please keep this discussion to the original question though, or open a new one if you need help for your own case. [On accidentally hijacking issues](https://github.com/nextauthjs/next-auth/issues/419)

**Reply by iamchathu** [reply]
@balazsorban44 https://codesandbox.io/s/eager-sun-7ly2e?file=/pages/index.tsx This is the minimal setup.

**Reply by balazsorban44** [reply]
It looks like you forgot to send the jwt token into `session`

https://next-auth.js.org/configuration/callbacks#session-callback

Check out the session callbacks second parameter. I think your access token should be there. :)

**Reply by iamchathu** [reply]
@balazsorban44 I tried it as below but accessToken is undefined there too.

```ts
 jwt: async (token, _user, account, _profile, _isNewUser) => {
      const result = { ...token, accessToken: account?.accessToken };
      return Promise.resolve(result);
    },
    session: async (session, user) => {
      console.log('USER : ', user);
      return Promise.resolve(session);
    },
```


Output is 

```bash
USER :  {
  name: 'xxxxx',
  email: 'xxxxxx@xxxxxx.xx',
  picture: 'https://s.gravatar.com/avatar/xxxxx.png',
  iat: 1607485526,
  exp: 1610077526,
  accessToken: undefined
}

```

**Comment by balazsorban44** [comment, marked as answer]
@iamchathu 

https://codesandbox.io/s/quizzical-wave-phrfj?file=/pages/api/auth/%5B...nextauth%5D.ts

Here is a working CodeSandbox for you. :)

The problem was that you did not check if `account` existed in `jwt`. I understood your intention with using optional chaining, but remember, it will just return undefined for `account` whenever `jwt` re-runs, resetting your `accessToken` you set in the first run.

I also added the `accessToken` to the session with the help of the `session` callback.

I created an Auth0 client and tested it, so I am sure it should work!

**Reply by iamchathu** [reply]
@balazsorban44 Thank you very much. I have updated my code and CSB too. But when I try to access the `accessToken` using `useSession` or getSession it returns `undefined`. I checked logging out token in callback and seems it sets accessToken without issues now.

**Reply by balazsorban44** [reply]
I'm not sure I understand. The problem is solved now, right? Or do you still need help?

**Reply by iamchathu** [reply]
I still can not access the accessToken value from useSession or getSession methods. Even though we have successfully setup the callbacks. I have updated the codesandbox with this fix and more code with useSession.

https://codesandbox.io/s/eager-sun-7ly2e?file=/pages/index.tsx

**Reply by balazsorban44** [reply]
Please have another look at my CodeSandbox and yours. You are still missing the session callback.

**Reply by iamchathu** [reply]
Sorry I just saw that. Seems there is an issue with TypeScript Types. I'm getting following error when Building.

```bash

Type error: Property 'accessToken' does not exist on type 'User'.

  31 |     },
  32 |     session: async (session, user) => {
> 33 |       session.accessToken = user.accessToken;
     |                                  ^
  34 | 
  35 |       return Promise.resolve(session);
  36 |     },
```

**Reply by iamchathu** [reply]
Also according to Auth0 [documentation](https://auth0.com/docs/tokens/access-tokens/validate-access-tokens) accessToken should be a JWT token. But current token I'm getting is not a valid JWT. (No dot divided 2 parts) ex: kRcsKWAT6XajGTsGws7fdJiBsX89BNz9

**Reply by iamchathu** [reply]
According to the Auth0 docs the token we getting is an opaque token.

`
How do I request a JWT?
In order to receive a JWT you must include an audience parameter with your token request. Typically, this would be an external API, like a custom API you have registered in the dashboard. See this doc for details:`

https://auth0.com/docs/tokens/access-tokens/get-access-tokens

**Reply by balazsorban44** [reply]
As far as I know, we just forward the auth token as is. I recommend you trying the same flow in postman, and check your access token. If you can verify that it looks different, I may be able to help further.

**Reply by iamchathu** [reply]
@balazsorban44 How about Types issue user.accessToken?. For opaque token issue We need to send Audience when we login in. But seems provider doesn't support that field too.

**Reply by balazsorban44** [reply]
We currently rely on the community to provide the types. Check out https://www.npmjs.com/package/@types/next-auth

There is an ongoing effort to rewrite next-auth in TS for better support at #516

**Reply by iamchathu** [reply]
I will work on types. How to set Audience in Auth0 provider?

**Reply by balazsorban44** [reply]
https://next-auth.js.org/configuration/providers#oauth-provider-options I think the `params` provider option could help with it

**Reply by DavidForDev** [reply]
@balazsorban44  hey, I have a new issue in 2023 I can't access to account object to get accessToken, so I lost a lot of days... I need help, do you can give me a plan on how to access the token? 

thank you
